clang/tools/clang-offload-deps/ClangOffloadDeps.cpp

```diff
@@ -0,0 +1,254 @@
+}
+
+int main(int argc, const char **argv) {
+  sys::PrintStackTraceOnErrorSignal(argv[0]);
+  ToolPath = sys::fs::getMainExecutable(argv[0], &ToolPath);
+
+  cl::HideUnrelatedOptions(ClangOffloadDepsCategory);
+  cl::SetVersionPrinter([](raw_ostream &OS) {
+    OS << clang::getClangToolFullVersion("clang-offload-deps") << '\n';
+  });
+  cl::ParseCommandLineOptions(
+      argc, argv,
+      "A tool for creating dependence bitcode files for offload targets. "
+      "Takes\nhost image as input and produces bitcode files, one per offload "
+      "target, with\nreferences to symbols that must be defined in target "
+      "images.\n");
+
+  if (Help) {
+    cl::PrintHelpMessage();
+    return 0;
+  }
+
+  // The number of output files and targets should match.
+  if (Targets.size() != Outputs.size()) {
+    reportError(
+        createStringError(errc::invalid_argument,
+                          "number of output files and targets should match"));
+    return 1;
+  }
+
+  // Verify that given targets are valid. Each target string is expected to have
+  // the following format
+  //     <kind>-<triple>
+  // where <kind> is host, openmp, hip, sycl or fpga,
+  // and <triple> is an offload target triple.
+  SmallVector<StringRef, 8u> Triples(Targets.size());
+  for (unsigned I = 0; I < Targets.size(); ++I) {
+    StringRef Kind;
+    std::tie(Kind, Triples[I]) = StringRef(Targets[I]).split('-');
+
+    bool KindIsValid = !Kind.empty() && StringSwitch<bool>(Kind)
+                                            .Case("host", true)
+                                            .Case("openmp", true)
+                                            .Case("hip", true)
+                                            .Case("sycl", true)
+                                            .Case("fpga", true)
+                                            .Default(false);
+
+    bool TripleIsValid = !Triples[I].empty() &&
+                         Triple(Triples[I]).getArch() != Triple::UnknownArch;
+
+    if (!KindIsValid || !TripleIsValid) {
+      SmallVector<char, 128u> Buf;
+      raw_svector_ostream Msg(Buf);
+      Msg << "invalid target '" << Targets[I] << "'";
+      if (!KindIsValid)
+        Msg << ", unknown offloading kind '" << Kind << "'";
+      if (!TripleIsValid)
+        Msg << ", unknown target triple '" << Triples[I] << "'";
+      reportError(createStringError(errc::invalid_argument, Msg.str()));
+      return 1;
+    }
+  }
+
+  // Read input file. It should have one of the supported object file formats.
+  Expected<OwningBinary<ObjectFile>> ObjectOrErr =
+      ObjectFile::createObjectFile(Input);
+  if (!ObjectOrErr) {
+    reportError(ObjectOrErr.takeError());
+    return 1;
+  }
+
+  // Then try to find a section in the input binary which contains offload
+  // symbol names and parse section contents.
+  DenseMap<StringRef, SmallDenseSet<StringRef>> Target2Symbols;
+  for (SectionRef Section : ObjectOrErr->getBinary()->sections()) {
```

Comment:
This seems to implement `std::find_if` logic, so I would use standard library function instead of for-loop.

Reply:
It is indeed close to `find_if`, but using `find_if` here would be problematic because predicate would be required to handle `Error` which can be returned by `Section.getName()` call (see line 148).

Reply:
Okay. Let's ignore this suggestion.